### DISCUSSION TITLE
Markup.Declarative C# Composition Syntax - Attaching Behaviors

### DISCUSSION THREAD

**Discussion by najak3d** [AvaloniaUI/Avalonia#7486] [Q&A] [answered]
First off, I think @gritsenko add-on for Markup.Declarative should have it's own category, perhaps.   It's a whole different method of using Avalonia, and raises it's own unique questions/concerns.

[https://github.com/AvaloniaUI/Avalonia.Markup.Declarative](https://github.com/AvaloniaUI/Avalonia.Markup.Declarative)

===
I'm trying to do some new things with this syntax, for which there is currently no Sample code.

1. Attaching CustomBehaviors, as shown by @wieslawsoltes package:
https://github.com/wieslawsoltes/AvaloniaBehaviors/tree/cdf92dbd1373bda70ab9dcbada009e17f2b0b798

2. Setting up a TreeView with Templates.

3. etc.

More sample could would be really awesome here, since 99% of the Avalonia documentation/samples are all for the AXAML syntax.

We find the C# notation approach highly attractive, as do many others.   Flutter/Dart have gained immense popularity very quickly much based on the concept of easy Code-UI composition (not XAML).

Promoting this Add-On C# syntax will attract MANY to Avalonia, and prevent those who would prefer C#/.NET from jumping ship to Flutter/Dart (which many projects are starting to do, since Xamarin.Form suck).

**Comment by wieslawsoltes** [comment, marked as answer]
As for setting custom behaviors from code:
```XAML
  <Grid RowDefinitions="Auto,Auto">
    <TextBox Name="TextBox" Text="Hello" Grid.Row="0" Margin="5" />
    <Button Name="ChangeProperty" Content="Change Property" Grid.Row="1" Margin="5,0,5,5">
      <i:Interaction.Behaviors>
        <ia:EventTriggerBehavior EventName="Click" SourceObject="{Binding #ChangeProperty}">
          <ia:ChangePropertyAction TargetObject="{Binding #TextBox}" PropertyName="Text" Value="World" />
        </ia:EventTriggerBehavior>
      </i:Interaction.Behaviors>
    </Button>
  </Grid>
```

You can do something like that (probably betters ways to do exists) for start and then refactor as fluent api or extensions methods:
```C#
    private void SetBehaviors()
    {
        var changeProperty = this.Find<Button>("ChangeProperty");
        var textBox = this.Find<TextBox>("TextBox");

        var eventTriggerBehavior = new EventTriggerBehavior()
        {
            EventName = "Click",
            SourceObject = changeProperty
        };

        var changePropertyAction = new ChangePropertyAction(textBox, "Text", "World");

        eventTriggerBehavior.Actions?.Add(changePropertyAction);

        var behaviors = new BehaviorCollection()
        {
            eventTriggerBehavior
        };

        Avalonia.Xaml.Interactivity.Interaction.SetBehaviors(changeProperty, behaviors);
    }
```

**Comment by wieslawsoltes** [comment]
As for using Avalonia from code I recommend looking at unit tests

TreeView https://github.com/AvaloniaUI/Avalonia/blob/master/tests/Avalonia.Controls.UnitTests/TreeViewTests.cs

ListBox https://github.com/AvaloniaUI/Avalonia/blob/master/tests/Avalonia.Controls.UnitTests/ListBoxTests.cs
